src/critnib/critnib.c

```diff
@@ -472,13 +555,90 @@ void *critnib_remove(struct critnib *c, word key) {
 
 del_leaf:
     value = k->value;
+    utils_atomic_store_release_ptr(&k->to_be_freed, value);
+    utils_atomic_store_release_ptr(&k->value, NULL);
     c->pending_del_leaves[del] = k;
+    *ref = k;
 
 not_found:
     utils_mutex_unlock(&c->mutex);
     return value;
 }
 
+/*
+ * critnib_release -- release a reference to a key
+ */
+int critnib_release(struct critnib *c, void *ref) {
+    if (!c || !ref) {
+        return -1;
+    }
+
+    struct critnib_leaf *k = (struct critnib_leaf *)ref;
+
+    uint64_t ref_count;
+    utils_atomic_load_acquire_u64(&k->ref_count, &ref_count);
+
+    if (ref_count == 0) {
+        return -1;
+    }
+
+    /* decrement the reference count */
+    if (utils_atomic_decrement_u64(&k->ref_count) == 0) {
+        void *to_be_freed = NULL;
+        utils_atomic_load_acquire_ptr(&k->to_be_freed, &to_be_freed);
+        if (to_be_freed) {
+            utils_atomic_store_release_ptr(&k->to_be_freed, NULL);
+            if (c->cb_free_leaf) {
+                c->cb_free_leaf(c->leaf_allocator, to_be_freed);
+            }
+        }
+    }
+
+#ifndef NDEBUG
+    // check if the reference count is overflowed
+    utils_atomic_load_acquire_u64(&k->ref_count, &ref_count);
+    assert((ref_count & (1ULL << 63)) == 0);
+    assert(ref_count != (uint64_t)(0 - 1ULL));
+#endif
+
+    return 0;
+}
+
+/*
+ * critnib_remove_release -- delete a key from the critnib structure
+ *
+ * Returns 0 on success, -1 if the key was not found.
+ */
+int critnib_remove_release(struct critnib *c, word key) {
+    void *ref;
+    void *value = critnib_remove(c, key, &ref);
+    if (ref) {
+        critnib_release(c, ref);
+    }
+
+    return (value) ? 0 : -1;
+}
+
+/*
+ * Check if the leaf has just been removed (i.e. if ref_count == 0).
+ * If so, we return -1 (failure), otherwise increment the reference
+ * counter and return 0 (success).
+ */
+static inline int increment_ref_count(struct critnib_leaf *k) {
+    uint64_t ref_count;
+    utils_atomic_load_acquire_u64(&k->ref_count, &ref_count);
+    if (ref_count == 0) {
+        return -1;
+    }
+
+    if (utils_atomic_increment_u64(&k->ref_count) == 1) {
+        utils_atomic_decrement_u64(&k->ref_count);
+        return -1;
+    }
```

Comment:
Why we check twice if ref_count is zero? 


also isn't it a race, between increment and decrement? If between both operation we have very long sleep, this node might be all ready reused, and count be different.

Reply:
`ref_count` can be incremented only if `ref_count > 0` - that's why the first check is needed.
After having incremented `ref_count`, we have to check if `ref_count == 1` to make sure it wasn't equal 0 before incrementing. If `ref_count == 1` we have to decrement it and exit. There is no other way to handle it.

Reply:
So why we have first check second check anyway

Reply:
Also you should use compare and swap instead increment and decrement to eliminate  race

Reply:
Do you mean like that?
```c
static inline int increment_ref_count(struct critnib_leaf *k) {
    uint64_t expected;
    uint64_t desired;

    do {
        utils_atomic_load_acquire_u64(&k->ref_count, &expected);
        if (expected == 0) {
            return -1;
        }
        desired = expected + 1;
    } while (!utils_compare_exchange_u64(&k->ref_count, &expected, &desired));

    return 0;
}
```

Reply:
Done